In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd 
import csv
import re
import datetime
pd.set_option('display.max_columns', None)

# 1. Đọc dữ liệu từ file
Đọc từ file lên đã rồi làm gì thì làm :D 

In [2]:
user_df = pd.read_csv("data/user.csv")
playlist_df = pd.read_csv("data/playlist.csv")
track_df = pd.read_csv("data/track.csv")

In ra phần head của dữ liệu, lướt sơ qua và hiểu thông tin các cột.

In [3]:
user_df.head()

avatar_url     city  comments_count  \
0  https://i1.sndcdn.com/avatars-000012778523-59q...   Berlin             463   
1  https://i1.sndcdn.com/avatars-000000018772-c5d...    Tokyo               0   
2  https://i1.sndcdn.com/avatars-000000018772-c5d...    Tokyo               0   
3  https://i1.sndcdn.com/avatars-N3HNyXVCJnJ8YiLQ...  Tampere             241   
4  https://i1.sndcdn.com/avatars-000000012820-990...      NaN               0   

  country_code            created_at  \
0           DE  2007-09-22T14:21:00Z   
1           JP  2008-03-06T12:56:42Z   
2           JP  2008-03-06T12:56:42Z   
3           FI  2008-04-19T17:43:00Z   
4          NaN  2008-03-08T00:32:42Z   

                            creator_subscriptions  \
0                   [{'product': {'id': 'free'}}]   
1                   [{'product': {'id': 'free'}}]   
2                   [{'product': {'id': 'free'}}]   
3  [{'product': {'id': 'creator-pro-unlimited'}}]   
4                   [{'product': {'id': 'free'}}]   

                           creator_subscription  \
0                   {'product': {'id': 'free'}}   
1                   {'product': {'id': 'free'}}   
2                   {'product': {'id': 'free'}}   
3  {'product': {'id': 'creator-pro-unlimited'}}   
4                   {'product': {'id': 'free'}}   

                                         description  followers_count  \
0                        Artist & Founder SoundCloud           133673   
1  Floor Limit is the music label in Japan.We hav...              265   
2  Floor Limit is the music label in Japan.We hav...              265   
3  I <3 synths. 1/2 of U&K. Host of Etkoplasma ra...             1502   
4  Final Headphone Science album 'Frozen Love Amo...              877   

   followings_count first_name       full_name  groups_count    id  kind  \
0              1270       Eric  Eric Wahlforss             0   183  user   
1                 5        Ark      Ark Kozuki             0  1691  user   
2                 5        Ark      Ark Kozuki             0  1691  user   
3              2002        Kim             Kim             0  2157  user   
4                 6        NaN             NaN             0  1719  user   

          last_modified  last_name  likes_count  playlist_likes_count  \
0  2020-12-16T21:40:33Z  Wahlforss          306                    10   
1  2020-08-17T12:33:59Z     Kozuki            3                     0   
2  2020-08-17T12:33:59Z     Kozuki            3                     0   
3  2021-08-17T10:55:32Z        NaN          826                    54   
4  2017-09-07T21:15:46Z        NaN            1                     0   

          permalink                            permalink_url  playlist_count  \
0             forss             https://soundcloud.com/forss               4   
1               ark               https://soundcloud.com/ark               3   
2               ark               https://soundcloud.com/ark               3   
3             kimik             https://soundcloud.com/kimik               5   
4  headphonescience  https://soundcloud.com/headphonescience               4   

   reposts_count  track_count                                    uri  \
0            NaN           26   https://api.soundcloud.com/users/183   
1            NaN            5  https://api.soundcloud.com/users/1691   
2            NaN            5  https://api.soundcloud.com/users/1691   
3            NaN           73  https://api.soundcloud.com/users/2157   
4            NaN           46  https://api.soundcloud.com/users/1719   

                     urn           username  verified  \
0   soundcloud:users:183              Forss      True   
1  soundcloud:users:1691                ark     False   
2  soundcloud:users:1691                ark     False   
3  soundcloud:users:2157              Kimik     False   
4  soundcloud:users:1719  Headphone Science     False   

                                             visuals  \
0  {'urn': 'soundcloud:users:183', 'enab

In [4]:
playlist_df.head()

artwork_url            created_at  \
0  https://i1.sndcdn.com/artworks-000067273270-ui...  2008-04-30T01:21:15Z   
1                                                NaN  2008-05-01T04:12:43Z   
2                                                NaN  2008-05-01T04:13:55Z   
3                                                NaN  2008-05-01T17:31:06Z   
4                                                NaN  2008-05-05T03:06:57Z   

                                         description  duration embeddable_by  \
0  My 2003 debut album made on the road with a la...   3247119           all   
1                                     artist / noe\n    354742           all   
2                                                NaN    254066           all   
3                                                NaN  50620378           all   
4        songs i made for a rave i played at in 2007   2498158           all   

              genre  id      kind      label_name         last_modified  \
0        Electronic  18  playlist  sonarkollektiv  2021-07-21T10:53:42Z   
1       ambient pop  32  playlist     Floor Limit  2015-10-13T13:39:25Z   
2  abstract hip-hop  33  playlist     Floor Limit  2015-10-13T13:35:10Z   
3               NaN  34  playlist             NaN  2021-10-16T14:55:49Z   
4     smoke & chill  38  playlist             NaN  2015-10-13T13:39:25Z   

               license  likes_count  managed_by_feeds       permalink  \
0  all-rights-reserved         1544             False        soulhack   
1  all-rights-reserved            2             False             noe   
2  all-rights-reserved            1             False      akt-the-jn   
3  all-rights-reserved            1             False        dj-mixes   
4  all-rights-reserved            3             False  the-rave-songs   

                                       permalink_url  public purchase_title  \
0         https://soundcloud.com/forss/sets/soulhack    True            NaN   
1                https://soundcloud.com/ark/sets/noe    True            NaN   
2         https://soundcloud.com/ark/sets/akt-the-jn    True            NaN   
3         https://soundcloud.com/kimik/sets/dj-mixes    True            NaN   
4  https://soundcloud.com/headphonescience/sets/t...    True            NaN   

                                        purchase_url          release_date  \
0     http://www.sonarkollektiv.com/releases/SK006CD  2003-06-02T00:00:00Z   
1  http://www.amazon.co.jp/beautiful-world-noe/dp...  2005-07-30T00:00:00Z   
2                                                NaN  2005-06-29T00:00:00Z   
3                                                NaN                   NaN   
4                                                NaN  2007-01-01T00:00:00Z   

   reposts_count  secret_token sharing  \
0            262           NaN  public   
1              0           NaN  public   
2              0           NaN  public   
3              0           NaN  public   
4              1           NaN  public   

                                            tag_list  \
0                           downtempo sampling beats   
1  ambient electronica IDM electropop japan "girl...   
2                      "hip hop" abstract breakbeats   
3                                                NaN   
4                                                NaN   

                                  title  \
0                              Soulhack   
1               noe - a beautiful world   
2  AKT THE JN - MIDNIGHT RAVERS RIDE ON   
3                              Dj Mixes   
4                        the rave songs   

                                       uri  user_id set_type  is_album  \
0  https://api.soundcloud.com/playlists/18      183    album      True   
1  https://api.soundcloud.com/playlists/32     1691    album      True   
2  https://api.soundcloud.com/playlists/33     1691    album      True   
3  https://api.soundcloud.com/playlists/34     2157      NaN     False   
4  https://api.soundcloud.com/playlist

In [5]:
track_df.head()

artwork_url caption  commentable  \
0  https://i1.sndcdn.com/artworks-000067273298-mj...     NaN         True   
1  https://i1.sndcdn.com/artworks-000067273301-we...     NaN         True   
2  https://i1.sndcdn.com/artworks-000067273305-sn...     NaN         True   
3  https://i1.sndcdn.com/artworks-000067273316-sm...     NaN         True   
4  https://i1.sndcdn.com/artworks-000067273321-fa...     NaN         True   

   comment_count            created_at  \
0          104.0  2007-09-22T14:24:38Z   
1          176.0  2007-09-22T14:32:59Z   
2           68.0  2007-09-22T14:41:54Z   
3          613.0  2007-09-22T14:45:46Z   
4           31.0  2007-09-22T15:00:17Z   

                                         description  downloadable  \
0  Intro from the Soulhack Album, originally crea...          True   
1                            From the Soulhack album          True   
2                            From the Soulhack album          True   
3  From the Soulhack album,&nbsp;recently feature...          True   
4                            From the Soulhack album          True   

   download_count  duration  full_duration embeddable_by       genre  \
0          6531.0    142236         142236           all  Electronic   
1          7288.0    274442         274442           all  Electronic   
2          3942.0    463908         463908           all  Electronic   
3         17514.0    213890         213890           all  Electronic   
4          3134.0    291395         291395           all  Electronic   

   has_downloads_left   id   kind      label_name         last_modified  \
0               False  290  track  sonarkollektiv  2021-01-05T06:42:33Z   
1               False  291  track  sonarkollektiv  2020-03-19T06:15:34Z   
2               False  292  track  sonarkollektiv  2020-03-19T06:08:42Z   
3               False  293  track  sonarkollektiv  2021-11-29T17:04:11Z   
4               False  294  track  sonarkollektiv  2019-06-12T15:07:06Z   

               license  likes_count       permalink  \
0  all-rights-reserved        332.0      city-ports   
1  all-rights-reserved       1068.0        soulhack   
2  all-rights-reserved        596.0  funk-for-nerds   
3  all-rights-reserved       2570.0     flickermood   
4  all-rights-reserved        261.0  using-splashes   

                                 permalink_url  playback_count  public  \
0      https://soundcloud.com/forss/city-ports        260313.0    True   
1        https://soundcloud.com/forss/soulhack        286623.0    True   
2  https://soundcloud.com/forss/funk-for-nerds        234596.0    True   
3     https://soundcloud.com/forss/flickermood        947286.0    True   
4  https://soundcloud.com/forss/using-splashes        202039.0    True   

                                  publisher_metadata purchase_title  \
0  {'id': 290, 'urn': 'soundcloud:tracks:290', 'a...            NaN   
1  {'id': 291, 'urn': 'soundcloud:tracks:291', 'a...            NaN   
2  {'id': 292, 'urn': 'soundcloud:tracks:292', 'a...            NaN   
3  {'id': 293, 'urn': 'soundcloud:tracks:293', 'a...            NaN   
4  {'id': 294, 'urn': 'soundcloud:tracks:294', 'a...            NaN   

                                       purchase_url          release_date  \
0  http://sonarkollektiv.com/tracks/DE-P96-03-00040  2003-06-02T00:00:00Z   
1  http://sonarkollektiv.com/tracks/DE-P96-02-01001  2003-06-02T00:00:00Z   
2  http://sonarkollektiv.com/tracks/DE-P96-03-00041  2003-06-02T00:00:00Z   
3  http://sonarkollektiv.com/tracks/DE-P96-03-00042  2003-06-02T00:00:00Z   
4  http://sonarkollektiv.com/tracks/DE-P96-03-00002  2003-06-02T00:00:00Z   

   reposts_count  secret_token sharing     state  streamable         tag_list  \
0             31           NaN  public  finished        True  cinematic intro   
1             84           NaN  public  finished        True             jazz   
2             62           NaN  public  finished        True             jazz   
3            400           NaN

Ta thấy rằng cột đầu tiên chứa url dẫn tới hình ảnh avatar, artwork,... là không cần thiết trong việc phân tích dữ liệu (trừ khi ta phân tích dữ liệu ảnh, điều mà sẽ không diễn ra trong đồ án này).  
Vì vậy, phần cột dữ liệu ấy sẽ được bỏ đi đầu tiên.
**Lưu ý : chỉ chạy ô bên dưới 1 lần duy nhất, vì khi chạy lần thứ 2, cột cần xóa sẽ không tồn tại trong dataframe nữa, khi đó hàm xóa không tìm thấy cốt cần xóa sẽ báo lỗi**

In [6]:
try:
    track_df = track_df.drop(columns=['artwork_url'])
except:
    pass
try:
    user_df = user_df.drop(columns=['avatar_url'])
except:
    pass
try:
    playlist_df = playlist_df.drop(columns=['artwork_url'])
except:
    pass

Điểm qua xem có bao nhiều dòng ở mỗi tập dữ liệu :

In [7]:
print("Số dòng của track : {}".format(len(track_df)))
print("Số dòng của play list : {}".format(len(playlist_df)))
print("Số dòng của user : {}".format(len(user_df)))

Số dòng của track : 4854
Số dòng của play list : 1011
Số dòng của user : 1011


# 2. Xác định cột khóa chính, loại bỏ cột không cần thiết và loại bỏ dòng trùng
Để loại bỏ những dòng bị trùng, đầu tiên ta phải xác định cột ID/khóa chính (cột định danh của mỗi bảng dữ liệu), vì đó là cơ sở để xem xét các dòng có bị trùng không.

## 2.1. Tập dữ liệu user
- Với tập dữ liệu user, ta thấy có 3 trường là `uri`,`urn`,`username`,`permalink_url`,`id` là ba cột nếu xét trên ý nghĩa thực tế là ba cột khóa chính (tức mỗi record sẽ có 1 giá trị phân biệt ở cột này).Ba cột đều có thể trở thành khóa chính.  
- Tuy nhiên, khi xét trên công việc phân tích dữ liệu, cột `uri` và `urn` không mang ý nghĩa lắm vì 2 trường đó là url của API.  
- Cột `permalink_url` là cột có thể suy ra từ `username`, hay nói cách khác từ `username` có thể suy dẫn ra `permalink_url`, điều này sẽ khiến cho thông tin bị trùng lắp, dư thừa và không hiệu quả khi đưa vào mô hình để phân tích dữ liệu. Thực tế, ta có thể phát hiện việc này qua mô hình xác suất (ở đây ta phát hiện bằng cách dò bằng mắt). Tuy nhiên, độ lớn của đồ án chưa đủ để chúng ta cần dùng đến mô hình ấy :).  
- Cột `user_name` có vẻ cần thiết. Tuy nhiên, khi phân tích dữ liệu, khóa chính cần phải càng đơn giản càng tốt. Rõ ràng, `user_name` so với `id` thì không thể bằng trong việc định tính đối tượng/record. Về mặt thông tin, `user_name` cũng không thể so với cột `full_name` vì rõ ràng `full_name` cho cả họ lẫn tên, còn username chỉ là biệt danh, sẽ phân tích được gì từ dữ liêu biệt danh của người dùng chứ? Qua phân tích này, ta thấy cột `first_name` cũng khá thừa thãi nên sẽ loại bỏ nó luôn.
- Những cột như : `station_urn`, `station_permalink` đều suy ra được từ `id`
Vì vậy, ta để lại `id` làm khóa chính.

In [8]:
user_df = user_df.drop(columns=['uri','urn','permalink_url','station_permalink','permalink','username','station_urn'])

- Sau đó, ta tiến hành loại bỏ dòng trùng dựa vào cột `id`.
- Trước đó bảng user có 1011 dòng. 
- Ta sẽ xem xét sau khi loại bỏ dòng trùng dựa vào id thì còn bao nhiêu.
- Ngoài ra, ta sẽ so sánh khi loại bỏ dòng trùng khi dựa vào `id` với khi loại bỏ dòng trùng dựa vào dữ liệu của nguyên cột (tức tất cả các cột giống nhau hết mới xem là trùng) xem nhận xét `id` là khóa chính có đúng không.


In [9]:
print("Loại bỏ dòng trùng dựa vào cả dòng : {}".format(len(user_df.drop_duplicates())))
user_df = user_df.drop_duplicates(subset=['id'])
print("Loại bỏ dòng rùng dựa vào cột id :{}".format(len(user_df)))

Loại bỏ dòng trùng dựa vào cả dòng : 694
Loại bỏ dòng rùng dựa vào cột id :694


Như đã thấy, dòng `id` chính xác là khóa chính chúng ta cần tìm. Ta sẽ cho cột `id` index column trong dataframe để dễ xử lý cho các bước sau.

In [10]:
user_df =user_df.set_index('id')

## 2.2. Tập dữ liệu play list

Giống như khi làm với tập user, chúng ta cũng sẽ loại bỏ những dòng unique nhưng không cung cấp thông tin cần thiết cho việc phân tích dữ liệu.
- `id` : Đầu tiên, ta thấy cột `id` làm rất tốt việc định danh từng record một cách đơn giản là đánh số.
- `permalink`,`user`,`permalink_url` : Có thể cung cấp thông tin về user và định danh track. Tuy vậy, ta có thể dùng cột `user_id` để cung cấp thông tin về user sở hữu track hiệu quả hơn.
- `uri`:định danh không tốt bằng `id`
Bước cho cột `id` là khóa chính thực hiện giống như mục 2.1.

In [11]:
playlist_df = playlist_df.drop(columns = ['permalink','permalink_url','permalink','user','uri'])

In [12]:
print("Xóa cột trùng theo dòng : {}".format(len(playlist_df.drop_duplicates())))
playlist_df = playlist_df.drop_duplicates(['id'])
print("Xóa cột trùng theo id : {}".format(len(playlist_df)))
playlist_df = playlist_df.set_index('id')


Xóa cột trùng theo dòng : 1011
Xóa cột trùng theo id : 1011


## 2.3 Tập dữ liệu Track
Sẽ có 3 cột được loại bỏ:
- `permalink`,`permalink_url`,`uri`,`urn`,`user` : giống như những mục trên
Cột đưa ra lựa chọn khó nhất chính là `publisher_metadata` bởi ta không biết liệu ở công đoạn sau, người xử lý có cần những dữ liệu thể hiện các mối quan hệ giữa publisher và bài hát hay không. Cho nên quyết định cuối cùng là giữ lại cho chắc. 

In [13]:
track_df = track_df.drop(columns=['permalink','permalink_url','uri','urn','user'])

In [14]:
print("Xóa cột trùng theo dòng : {}".format(len(track_df.drop_duplicates())))
track_df = track_df.drop_duplicates(['id'])
print("Xóa cột trùng theo id : {}".format(len(track_df)))
track_df.set_index('id')

Xóa cột trùng theo dòng : 4732
Xóa cột trùng theo id : 4732


caption  commentable  comment_count            created_at  \
id                                                                 
290        NaN         True          104.0  2007-09-22T14:24:38Z   
291        NaN         True          176.0  2007-09-22T14:32:59Z   
292        NaN         True           68.0  2007-09-22T14:41:54Z   
293        NaN         True          613.0  2007-09-22T14:45:46Z   
294        NaN         True           31.0  2007-09-22T15:00:17Z   
...        ...          ...            ...                   ...   
111915     NaN         True            0.0  2009-01-14T23:45:40Z   
111973     NaN         True            0.0  2009-01-14T23:56:08Z   
112376     NaN         True            0.0  2009-01-15T05:12:35Z   
94313      NaN         True            2.0  2008-12-23T16:04:10Z   
98003      NaN         True            4.0  2008-12-31T00:26:07Z   

                                              description  downloadable  \
id                                                                        
290     Intro from the Soulhack Album, originally crea...          True   
291                               From the Soulhack album          True   
292                               From the Soulhack album          True   
293     From the Soulhack album,&nbsp;recently feature...          True   
294                               From the Soulhack album          True   
...                                                   ...           ...   
111915                                 dj set tek-minimal          True   
111973                                                NaN         False   
112376                                                NaN         False   
94313   Preversion of the Swiss-German Elektrorap Club...          True   
98003   A famous swiss german reggae singer fidgetly r...          True   

        download_count  duration  full_duration embeddable_by         genre  \
id                                                                            
290             6531.0    142236         142236           all    Electronic   
291             7288.0    274442         274442           all    Electronic   
292             3942.0    463908         463908           all    Electronic   
293            17514.0    213890         213890           all    Electronic   
294             3134.0    291395         291395           all    Electronic   
...                ...       ...            ...           ...           ...   
111915            19.0   4703999        4703999           all           NaN   
111973             0.0     75712          75712           all           NaN   
112376             0.0     53995          53995           all           NaN   
94313            129.0    250405         250405           all    Elektrorap   
98003            167.0    381622         381622           all  Fidget House   

        has_downloads_left   kind      label_name         last_modified  \
id                                                                        
290                  False  track  sonarkollektiv  2021-01-05T06:42:33Z   
291                  False  track  sonarkollektiv  2020-03-19T06:15:34Z   
292                  False  track  sonarkollektiv  2020-03-19T06:08:42Z   
293                  False  track  sonarkollektiv  2021-11-29T17:04:11Z   
294                  False  track  sonarkollektiv  2019-06-12T15:07:06Z   
...                    ...    ...             ...                   ...   
111915                True  track             NaN  2013-03-08T15:45:44Z   
111973               False  track             NaN  2009-05-08T19:29:36Z   
112376               False  track             NaN  2009-01-15T05:12:35Z   
94313                False  track   Inferno Muzik  2021-10-18T10:11:02Z   
98003                False  track             NaN  2021-10-18T10:11:02Z   

                    license  likes_count  playback_count  public  \
id                                                                 
290   

Như vậy, ta giảm được xấp xỉ 100 dòng so với ban đầu.

# 3. Xử lý giá trị thiếu
Với những cột thiếu dữ liệu, ta có hai hướng xử lý : 
- Hoặc bỏ cột ấy đi
- Hoặc xây dựng một mô hình suy diễn điền vào các gía trị thiếu. 
 
Với những cột có dữ liệu thiếu quá nhiều, vượt ngưỡng 50%, thì việc điền vào các giá trị thiếu bằng mô hình có thể sẽ không chính xác, ảnh hưởng đến các bước phân tích về sau nghiêm trọng.
Vì vậy, những cột có giá trị thiếu >50% đều sẽ bị loại bỏ!

In [15]:
percent = 50.0
min_count = int((100-percent)/100*playlist_df.shape[0]+1)
playlist_df = playlist_df.dropna(axis=1,thresh=min_count)
track_df = track_df.dropna(axis=1,thresh=min_count)
user_df = user_df.dropna(axis=1,thresh=min_count)

# 4. Chuyển đổi dữ liệu về kiểu phù hợp

## 4.1 Dữ liệu user
Đầu tiên ta xem sơ qua những kiểu dữ liệu của `user`

In [16]:
print(user_df.dtypes)
user_df.head()

city                     object
comments_count            int64
country_code             object
created_at               object
creator_subscriptions    object
creator_subscription     object
followers_count           int64
followings_count          int64
groups_count              int64
kind                     object
last_modified            object
likes_count               int64
playlist_likes_count      int64
playlist_count            int64
track_count               int64
verified                   bool
badges                   object
dtype: object


city  comments_count country_code            created_at  \
id                                                                  
183     Berlin             463           DE  2007-09-22T14:21:00Z   
1691     Tokyo               0           JP  2008-03-06T12:56:42Z   
2157   Tampere             241           FI  2008-04-19T17:43:00Z   
1719       NaN               0          NaN  2008-03-08T00:32:42Z   
2455  Borlänge               9           SE  2008-05-14T19:44:36Z   

                               creator_subscriptions  \
id                                                     
183                    [{'product': {'id': 'free'}}]   
1691                   [{'product': {'id': 'free'}}]   
2157  [{'product': {'id': 'creator-pro-unlimited'}}]   
1719                   [{'product': {'id': 'free'}}]   
2455                   [{'product': {'id': 'free'}}]   

                              creator_subscription  followers_count  \
id                                                                    
183                    {'product': {'id': 'free'}}           133673   
1691                   {'product': {'id': 'free'}}              265   
2157  {'product': {'id': 'creator-pro-unlimited'}}             1502   
1719                   {'product': {'id': 'free'}}              877   
2455                   {'product': {'id': 'free'}}              209   

      followings_count  groups_count  kind         last_modified  likes_count  \
id                                                                              
183               1270             0  user  2020-12-16T21:40:33Z          306   
1691                 5             0  user  2020-08-17T12:33:59Z            3   
2157              2002             0  user  2021-08-17T10:55:32Z          826   
1719                 6             0  user  2017-09-07T21:15:46Z            1   
2455                46             0  user  2015-10-18T13:05:36Z           33   

      playlist_likes_count  playlist_count  track_count  verified  \
id                                                                  
183                     10               4           26      True   
1691                     0               3            5     False   
2157                    54               5           73     False   
1719                     0               4           46     False   
2455                     5               4           29     False   

                                                 badges  
id                                                       
183   {'pro': False, 'pro_unlimited': False, 'verifi...  
1691  {'pro': False, 'pro_unlimited': False, 'verifi...  
2157  {'pro': False, 'pro_unlimited': True, 'verifie...  
1719  {'pro': False, 'pro_unlimited': False, 'verifi...  
2455  {'pro': False, 'pro_unlimited': False, 'verifi...

Ta nhận thấy dữ liệu cột `create_at`,`last_modified` có định dạng ngày tháng năm nhưng được biểu diễn dưới dạng object

In [17]:
user_df['created_at'] = pd.to_datetime(user_df['created_at'])
user_df['last_modified'] = pd.to_datetime(user_df['last_modified'])

## 4.2 Dữ liệu playlist
Đầu tiên ta xem sơ qua những kiểu dữ liệu của `playlist`

In [18]:
print(playlist_df.dtypes)
playlist_df.head()

created_at          object
description         object
duration             int64
embeddable_by       object
kind                object
last_modified       object
license             object
likes_count          int64
managed_by_feeds      bool
public                bool
reposts_count        int64
sharing             object
title               object
user_id              int64
is_album              bool
display_date        object
tracks              object
track_count          int64
dtype: object


created_at                                        description  \
id                                                                            
18  2008-04-30T01:21:15Z  My 2003 debut album made on the road with a la...   
32  2008-05-01T04:12:43Z                                     artist / noe\n   
33  2008-05-01T04:13:55Z                                                NaN   
34  2008-05-01T17:31:06Z                                                NaN   
38  2008-05-05T03:06:57Z        songs i made for a rave i played at in 2007   

    duration embeddable_by      kind         last_modified  \
id                                                           
18   3247119           all  playlist  2021-07-21T10:53:42Z   
32    354742           all  playlist  2015-10-13T13:39:25Z   
33    254066           all  playlist  2015-10-13T13:35:10Z   
34  50620378           all  playlist  2021-10-16T14:55:49Z   
38   2498158           all  playlist  2015-10-13T13:39:25Z   

                license  likes_count  managed_by_feeds  public  reposts_count  \
id                                                                              
18  all-rights-reserved         1544             False    True            262   
32  all-rights-reserved            2             False    True              0   
33  all-rights-reserved            1             False    True              0   
34  all-rights-reserved            1             False    True              0   
38  all-rights-reserved            3             False    True              1   

   sharing                                 title  user_id  is_album  \
id                                                                    
18  public                              Soulhack      183      True   
32  public               noe - a beautiful world     1691      True   
33  public  AKT THE JN - MIDNIGHT RAVERS RIDE ON     1691      True   
34  public                              Dj Mixes     2157     False   
38  public                        the rave songs     1719      True   

            display_date                                             tracks  \
id                                                                            
18  2008-04-30T01:21:15Z  [{'artwork_url': 'https://i1.sndcdn.com/artwor...   
32  2008-05-01T04:12:43Z                                                 []   
33  2008-05-01T04:13:55Z  [{'artwork_url': None, 'caption': None, 'comme...   
34  2008-05-01T17:31:06Z  [{'artwork_url': 'https://i1.sndcdn.com/artwor...   
38  2008-05-05T03:06:57Z  [{'artwork_url': None, 'caption': None, 'comme...   

    track_count  
id               
18           11  
32            0  
33            1  
34           10  
38            4

Ta thấy các cột `created_at`,`last_modified`,`display_date` cần được chuyển sang định dạng ngày tháng năm để dễ xử lý

In [19]:
playlist_df['created_at'] = pd.to_datetime(playlist_df['created_at'])
playlist_df['last_modified'] = pd.to_datetime(playlist_df['last_modified'])
playlist_df['display_date'] = pd.to_datetime(playlist_df['display_date'])

## 4.3 Dữ liệu track
Đầu tiên ta xem sơ qua những kiểu dữ liệu của `track`

In [20]:
print(track_df.dtypes)
track_df.head()

commentable               bool
comment_count          float64
created_at              object
description             object
downloadable              bool
download_count         float64
duration                 int64
full_duration            int64
embeddable_by           object
genre                   object
has_downloads_left        bool
id                       int64
kind                    object
label_name              object
last_modified           object
license                 object
likes_count            float64
playback_count         float64
public                    bool
publisher_metadata      object
purchase_url            object
release_date            object
reposts_count            int64
sharing                 object
state                   object
streamable                bool
tag_list                object
title                   object
track_format            object
user_id                  int64
waveform_url            object
display_date            object
media   

commentable  comment_count            created_at  \
0         True          104.0  2007-09-22T14:24:38Z   
1         True          176.0  2007-09-22T14:32:59Z   
2         True           68.0  2007-09-22T14:41:54Z   
3         True          613.0  2007-09-22T14:45:46Z   
4         True           31.0  2007-09-22T15:00:17Z   

                                         description  downloadable  \
0  Intro from the Soulhack Album, originally crea...          True   
1                            From the Soulhack album          True   
2                            From the Soulhack album          True   
3  From the Soulhack album,&nbsp;recently feature...          True   
4                            From the Soulhack album          True   

   download_count  duration  full_duration embeddable_by       genre  \
0          6531.0    142236         142236           all  Electronic   
1          7288.0    274442         274442           all  Electronic   
2          3942.0    463908         463908           all  Electronic   
3         17514.0    213890         213890           all  Electronic   
4          3134.0    291395         291395           all  Electronic   

   has_downloads_left   id   kind      label_name         last_modified  \
0               False  290  track  sonarkollektiv  2021-01-05T06:42:33Z   
1               False  291  track  sonarkollektiv  2020-03-19T06:15:34Z   
2               False  292  track  sonarkollektiv  2020-03-19T06:08:42Z   
3               False  293  track  sonarkollektiv  2021-11-29T17:04:11Z   
4               False  294  track  sonarkollektiv  2019-06-12T15:07:06Z   

               license  likes_count  playback_count  public  \
0  all-rights-reserved        332.0        260313.0    True   
1  all-rights-reserved       1068.0        286623.0    True   
2  all-rights-reserved        596.0        234596.0    True   
3  all-rights-reserved       2570.0        947286.0    True   
4  all-rights-reserved        261.0        202039.0    True   

                                  publisher_metadata  \
0  {'id': 290, 'urn': 'soundcloud:tracks:290', 'a...   
1  {'id': 291, 'urn': 'soundcloud:tracks:291', 'a...   
2  {'id': 292, 'urn': 'soundcloud:tracks:292', 'a...   
3  {'id': 293, 'urn': 'soundcloud:tracks:293', 'a...   
4  {'id': 294, 'urn': 'soundcloud:tracks:294', 'a...   

                                       purchase_url          release_date  \
0  http://sonarkollektiv.com/tracks/DE-P96-03-00040  2003-06-02T00:00:00Z   
1  http://sonarkollektiv.com/tracks/DE-P96-02-01001  2003-06-02T00:00:00Z   
2  http://sonarkollektiv.com/tracks/DE-P96-03-00041  2003-06-02T00:00:00Z   
3  http://sonarkollektiv.com/tracks/DE-P96-03-00042  2003-06-02T00:00:00Z   
4  http://sonarkollektiv.com/tracks/DE-P96-03-00002  2003-06-02T00:00:00Z   

   reposts_count sharing     state  streamable         tag_list  \
0             31  public  finished        True  cinematic intro   
1             84  public  finished        True             jazz   
2             62  public  finished        True             jazz   
3            400  public  finished        True        downtempo   
4             23  public  finished        True          samples   

            title  track_format  user_id  \
0      City Ports  single-track      183   
1        Soulhack  single-track      183   
2  Funk for Nerds  single-track      183   
3     Flickermood  single-track      183   
4  Using Splashes  single-track      183   

                                  waveform_url          display_date  \
0  https://wave.sndcdn.com/UiLhla0Ho0Ad_m.json  2007-09-22T14:24:38Z   
1  https://wave.sndcdn.com/AqFIRCAP7Arw_m.json  2007-09-22T14:32:59Z   
2  https://wave.sndcdn.com/EsdplDAYufuQ_m.json  2007-09-22T14:41:54Z   
3  https://wave.sndcdn.com/cWHNerOLlkUq_m.json  2007-09-22T14:45:46Z   
4  https://wave.sndcdn.com/hrNkZJd9Tjx4_m.json  2007-09-22T15:00:17Z   

                                               media  \
0  {'transcodings': [{'url': 'h

Có 2 vấn đề cần xử lý ở dữ liệu `user` : 
- Chuyển các cột `created_at`,`release_date` và `last_modified` sang dạng ngày tháng năm.
- Chuyển các cột có đáng ra có dữ liệu int nhưng lại mang dữ liệu float

In [21]:
# chuyển ngày làm tương tự như 2 tập dữ liệu trên
track_df['created_at'] = pd.to_datetime(track_df['created_at'])
track_df['last_modified'] = pd.to_datetime(track_df['last_modified'])
track_df['release_date'] = pd.to_datetime(track_df['release_date'])
track_df['display_date'] = pd.to_datetime(track_df['display_date'])

Xem các cột có dữ liệu float

In [22]:
dtypes = track_df.dtypes
dtypes
float_cols = (dtypes[(dtypes==np.float32) | (dtypes==np.float64)].index)
float_cols
# track_cols_float = set(track_dypte[(track_dypte==np.float_)])
# track_cols_float

Index(['comment_count', 'download_count', 'likes_count', 'playback_count'], dtype='object')

Ta thấy các dữ liệu này đều là biến đếm (đếm số lượng download, comment) và phải mang kiểu dữ liệu là Int. Tuy nhiên các biến này lại mang dữ liệu float, sẽ gây khó khăn khi thực hiện phân tích dữ liệu. Vì vậy ta sẽ chuyển chúng sang Int.

In [23]:
track_df[float_cols] = track_df[float_cols].astype('Int64')